src/functions.rs

```diff
@@ -2827,23 +2895,99 @@ fn validate_argument<'a>(
             if arg_typeinfo.is_child_or_actual_type(target_argument.param_type, types) {
                 Ok(TypeInstance::new(&arg, arg_typeinfo, file, context))
             } else {
-                Err(ErrorItem::make_compile_or_internal_error(
-                    &format!(
-                        "Expected type inheriting {}",
-                        target_argument.param_type.name
-                    ),
+                Err(validate_argument_error_handler(
+                    arg,
+                    arg_typeinfo,
+                    target_argument,
+                    types,
                     file,
-                    arg.get_range(),
-                    &format!(
-                        "This type should inherit {}",
-                        target_argument.param_type.name
-                    ),
+                    func_info,
                 ))
             }
         }
     }
 }
 
+// Helper function for handling the error case for validate_argument.
+fn validate_argument_error_handler(
+    arg: ArgForValidation,
+    arg_typeinfo: &TypeInfo,
+    target_argument: &FunctionArgument,
+    types: &TypeMap,
+    file: Option<&SimpleFile<String, String>>,
+    func_info: Option<&FunctionInfo>,
+) -> ErrorItem {
+    // If there is no function call or the call is not an asociated one fall through and return
+    // the "standard" error.
+    // If this is the associated call we need to do some more digging to give the user a better
+    // error message.
+    // Unwraps of func_info are safe in this block because of the false return on map_or
+    if func_info.map_or(false, |f| f.is_associated_call) {
+        let mut error = ErrorItem::make_compile_or_internal_error(
+            &format!(
+                "Expected type inheriting {} for associated call",
+                target_argument.param_type.name
+            ),
+            arg_typeinfo.get_file().as_ref(),
+            arg_typeinfo.name.get_range(),
+            &format!(
+                "An associated call: '{}' was made for this domain.  That call requires this domain inherit {}",
+                func_info.unwrap().name, target_argument.param_type.name
+            ),
+        );
+        // We don't know which class the associated call came from, but that would
+        // be nice information for the user.  So we search through all the associate class
+        // annotations, and find the class we are looking for. Then add that to the
+        // error message.
+        for annotation in arg_typeinfo.get_annotations() {
```

Comment:
Can we not point at the associated_call itself?  We're validating it right now, so we should have everything we need, right?  Would that make for a clearer error message?

Reply:
If you are talking about pulling it from function info the, it has a declaration file of: `SimpleFile { name: "", source: "", line_starts: [0] }`

Reply:
Oh, right.  If the particular @associated call in this instance is derived, it wouldn't have a valid file.  What this PR does is find the associated resource, but get the name of its parent.

Does this work with a nested syntax like below?

```
domain dom {
    resource res {
        @associated_call
        fn call(userdomain source) {
            allow(source, this, file, read);
        }
    }
}
```

Reply:
```
domain dom {
    resource res {
        @associated_call
        fn call(userdomain source) {
            allow(source, this, file, read);
        }
    }
}
```
When added to assocaite.cas gives the error:
```
error: Expected type inheriting userdomain for associated call
    ┌─ data/policies/associate.cas:118:8
    │
118 │ domain dom {
    │        ^^^ An associated call: 'call' was made for this domain.  That call requires this domain inherit userdomain
```

```
domain dom inherits userdomain {
    resource res {
        @associated_call
        fn call(userdomain source) {
            allow(source, this, file, read);
        }
    }
}
```
Compiles cleanly.

Reply:
Okay, I've been staring at this bewildered for a while, and I think I figured out what's going on.  I think it's because of this line:

https://github.com/dburgener/cascade/blob/5daa36d1e4c455f4108fcc634f1c3ff175c36f65/src/compile.rs#L141

When we create a nested resource, we make a fake parent for it, and that parent is created from a clone of the nested TypeDecl.  That fake parent ends up in this case having the name `res`, while the real associated resource is `dom.res`.  So strictly, the declaration location of `res` shouldn't exist and the declaration location of `dom.res` is the nested declaration.  However, due to the way `res` is made from `dom.res`, they end up pointing at the same location.

The problem here, as @pebenito pointed out, is that if there is a collection named `res` in the same policy, we have an expected conflict.  In order to address this, I'm thinking that I'll want to declare `dom.res` directly in extend_type_map, rather than declaring a fake parent to associate.  I believe that that expected future change will break this.

I *think* the most correct thing to do is to check the file and range on the FunctionInfo we're working on now.  If it's valid, we can add a message and return.  If it's not valid, then that means this \@associated_call was derived.  So we need to locate the parents of func_info.class, and check their \@associated_calls, until we find one that both 1. Has a valid file and range, and 2. Has the type signature we're currently complaining about.  That should theoretically always exist, since if this \@associated_call isn't declared in source, it has to have been derived from somewhere.

Reply:
In the newest push the error now looks like:
```
error: Expected type inheriting userdomain for associated call
   ┌─ data/error_policies/bad_inherits_associated_call.cas:11:8
   │
11 │ domain bad_userdomain {
   │        ^^^^^^^^^^^^^^ An associated call: 'associated_call_from_user_tmp' was made for this domain.  That call requires this domain inherit userdomain
   │
   ┌─ data/error_policies/bad_inherits_associated_call.cas:5:5
   │
 5 │     fn associated_call_from_user_tmp(userdomain source) {
   │        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Associated function found here
```